crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,171 @@
+    pub fn durability(&self) -> Option<u32> { None }
+    pub fn stack_size(&self) -> u32         { 64 }
+}
+
+/// Represents an item stack.
+///
+/// An item stack includes an item type, an amount and a bunch o properties (enchantments, etc.)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ItemStack {
+
+    // TODO Migrate all the codegen boilerplate.
+    /// The item type of this `ItemStack`.
+    pub item: Item,
+
+    /// The number of items in the `ItemStack`.
+    pub count: u32,
+
+    /// Damage to the item, if it's damageable.
+    pub damage: Option<u32>,
+
+    /// Repair cost of the item, if it's repairable.
+    pub repair_cost: Option<u32>,
+
+    /// Enchantments applied to the item.
+    pub enchantments: Vec<Enchantment>,
+
+    // TODO Add other properties like title and lore.
+}
+
+impl ItemStack {
+
+    /// Creates a new `ItemStack`
+    pub fn new(item: Item, count: u32) -> Self {
+        Self {
+            item,
+            count,
+            damage: item.durability().map(|_| 0)
+        }
+    }
+
+    /// Returns whether the given item stack has
+    /// the same type as (but not necessarily the same
+    /// amount as) `self`.
+    pub fn has_same_type(&self, other: &Self) -> bool {
+        self.item == other.item
+    }
+
+    /// Returns whether the given item stack has the same damage
+    /// as `self`.
+    pub fn has_same_damage(&self, other: &Self) -> bool {
+        self.damage == other.damage
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same count as (but not necessarily the same
+    /// type as) `self`.
+    pub fn has_same_count(&self, other: &Self) -> bool {
+        self.count == self.count
+    }
+
+    /// Returns whether the given `ItemStack` has the same
+    /// type and count as `self`.
+    pub fn has_same_type_and_count(&self, other: &Self) -> bool {
+        self.item == other.item && self.count == other.count
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same type and damage as `self`.
+    pub fn has_same_type_and_damage(&self, other: &Self) -> bool {
+        self.item == other.item && self.damage == other.damage
+    }
+
+    /// Returns the item type for this `ItemStack`.
+    pub fn item(&self) -> Item {
+        self.item
+    }
+
+    /// Returns the number of items in this `ItemStack`.
+    pub fn count(&self) -> u32 {
+        self.count
+    }
+
+    /// Adds more items to this `ItemStack`. Returns the new count.
+    pub fn add(&mut self, count: u32) -> u32 {
+        self.count += count;
+        self.cout
+    }
+
+    /// Removes some items from this `ItemStack`.
+    pub fn remove(&mut self, count: u32) -> Result<u32, ()> {
+        self.count = match self.count.checked_sub(count) {
+            Some(count) => count,
+            None => return Err(())
+        };
+        Ok(self.count)
+    }
+
+    /// Sets the item type for this `ItemStack`.
+    pub fn set_item(&mut self, item: Item) {
+        self.item = item

```

Comment:
Stack limited should be respected

Reply:
See comment about `set_count`. I would apply the same logic as the one applied there. If not, which behaviour should we implement? Return an `Err` or trim the count?

Reply:
Done, by returning `Err` variant when the stack size limits would be violated.